### PR TITLE
feat: instrument more rpc and upstream metrics

### PR DESCRIPTION
**Pull request by jonathankau**
# Description

Instruments basic metrics for monitoring inbound requests/responses and forwards to upstream nodes. Useful for tracking

## Type of change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 😎 New feature (non-breaking change which adds functionality)
- [ ] ⁉️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] ⚒️ Refactor (no functional changes)
- [ ] 📖 Documentation (updating or adding docs)

# How Has This Been Tested?

Ran with a config with 1 upstream pointed at https://rpc.ankr.com/eth. Ran a few curl commands and checked metrics:
```bash
curl --location --request POST 'localhost:8080' \
--header 'Content-Type: application/json' \
--data-raw '{
	"jsonrpc":"2.0",
	"method":"eth_blockNumber",
	"params":[],
	"id":1
}'


curl --location --request POST 'localhost:8080' \
--header 'Content-Type: application/json' \
--data-raw '{
	"jsonrpc":"2.0",
	"method":"dummy_method",
	"params":[],
	"id":1
}'
```

Screen Shot 2022-08-26 at 11 36 40 AM

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,7 @@
 # ⛩ node-gateway
+[![Test](https://github.com/satsuma-xyz/node-gateway/actions/workflows/test.yml/badge.svg)](https://github.com/satsuma-xyz/node-gateway/actions/workflows/test.yml)
```
Comment (brianluong) [inline README.md:2]: Think you may need to rebase off main against since I force pushed to amend a commit msg

File: internal/metrics/metrics.go [hunk 2]
```diff
@@ -18,12 +18,55 @@ const (
...
+			Name:      "rpc_response_size_bytes",
+			Help:      "Histogram of RPC response sizes.",
+			Buckets:   []float64{100, 500, 1000, 5000, 10000},
+		},
+		[]string{"code", "method"},
+	)
+
+	UpstreamRPCRequestsTotal = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamespace,
```
Comment (brianluong) [inline internal/metrics/metrics.go:55]: do these need `Subsystem?`

File: internal/route/router.go [hunk 3]
```diff
@@ -106,7 +107,24 @@ func (r *SimpleRouter) Route(requestBody jsonrpc.RequestBody) (*jsonrpc.Response
...
 	zap.L().Debug("Routing request to upstream.", zap.String("upstreamID", id), zap.Any("request", requestBody))
+	metrics.UpstreamRPCRequestsTotal.WithLabelValues(id, configToRoute.HTTPURL).Inc()
 
+	body, response, err := r.routeToConfig(requestBody, &configToRoute)
+
+	if err != nil {
+		metrics.UpstreamRPCRequestErrorsTotal.WithLabelValues(id, configToRoute.HTTPURL).Inc()
+	} else if body.Error != nil {
+		zap.L().Warn("Encountered error in upstream JSONRPC response.", zap.Any("request", requestBody), zap.Any("error", body.Error))
+		metrics.UpstreamRPCRequestErrorsTotal.WithLabelValues(id, configToRoute.HTTPURL).Inc()
```
Comment (brianluong) [inline internal/route/router.go:118]: What do you think about adding response code as a label?

Maybe `http-${statuscode}` and `jsonrpc-${statuscode}`?

File: internal/metrics/metrics.go [hunk 4]
```diff
@@ -18,12 +18,55 @@ const (
...
+			Subsystem: metricsSubsystem,
+			Name:      "rpc_requests_total",
+			Help:      "Count of total RPC requests.",
+		},
+		[]string{"code", "method"},
+	)
+
+	rpcRequestsDuration = promauto.NewHistogramVec(
```
Comment (brianluong) [inline internal/metrics/metrics.go:31]: Were you planning on saving these for a future PR?

Comment (brianluong) [reply internal/metrics/metrics.go:31]: The test in the PR description shows these metrics, so I'm assuming these automagically wired in somehow

Comment (jonathankau) [reply internal/metrics/metrics.go:31]: These are instrumented here: https://github.com/satsuma-xyz/node-gateway/pull/45/files#diff-6e56d7c057f6a96124b78f2b669dd7dface2718c304aece205017d114b3a7088R83

**Review by brianluong — APPROVED** [review]
🚢

**Pull request merged**